cmd/snap-update-ns/change.go

```diff
@@ -59,26 +60,141 @@ var changePerform func(*Change) ([]*Change, error)
 
 // changePerformImpl is the real implementation of Change.Perform
 func changePerformImpl(c *Change) ([]*Change, error) {
+	var changes []*Change
 	if c.Action == Mount {
+		// We may be asked to bind mount a file, bind mount a directory, mount
+		// a filesystem over a directory or create a symlink (which is abusing
+		// the "mount" concept slightly). That actual operation is performed in
+		// c.lowLevelPerform. Here we just set the stage to make that possible.
+		kind, _ := c.Entry.OptStr("x-snapd.kind")
+		path := c.Entry.Dir
+
+		// We use lstat to ensure that we don't follow a symlink in case one
+		// was set up by the snap. Note that at the time this is run, all the
+		// snap's processes are frozen.
+		fi, err := osLstat(path)
+
+		// In case we need to create something, some constants.
 		const (
 			mode = 0755
 			uid  = 0
 			gid  = 0
 		)
-		// Create target mount directory if needed.
-		if err := ensureMountPoint(c.Entry.Dir, mode, uid, gid); err != nil {
-			return nil, err
+
+		if err != nil && os.IsNotExist(err) {
+			// If the element doesn't exist we can attempt to create it.
+			// We will create the parent directory and then the final element
+			// relative to it. The traversed space may be writable so we try a
+			// to create the parent directory first.
+			switch kind {
+			case "":
+				err = secureMkdirAll(path, mode, uid, gid)
+			case "file":
+				err = secureMkfileAll(path, mode, uid, gid)
+			case "symlink":
+				path = filepath.Dir(c.Entry.Dir)
+				err = secureMkdirAll(path, mode, uid, gid)
+				if err == nil {
+					// Normally symlinks would be created
+					// later but calling c.lowLevelPerform
+					// now lets us check if the medium is
+					// writable and take advantage of the
+					// code below.
+					err = c.lowLevelPerform()
+				}
+			}
+
+			if err2, _ := err.(*ReadOnlyFsError); err2 != nil {
+				// If the writing failed because the underlying filesystem
+				// is read-only we can construct a writable mimic to fix that.
+				changes, err = createWritableMimic(err2.Path)
+				if err != nil {
+					return nil, fmt.Errorf("cannot create writable mimic over %q: %s", err2.Path, err)
+				}
+				// We can now try again.
+				switch kind {
```

Comment:
Can this go into a helper too? It looks almost identical as the switch block above (minus the path=... bit for symlink, which would be valid for second block and is just omited as was already set).